Replace 1-bit bitfields in RenderObjectBitfields with RenderObjectFlag

rniwa opened the pull request:
#### 9b9eb59716b8ed70b8ada714e1c03ee7cb0c7b23
<pre>
Replace 1-bit bitfields in RenderObjectBitfields with RenderObjectFlag
<a href="https://bugs.webkit.org/show_bug.cgi?id=266609">https://bugs.webkit.org/show_bug.cgi?id=266609</a>

Reviewed by Simon Fraser.

Replace macros to generate 1-bit bitfields in RenderObjectBitfields with newly introduced
RenderObjectFlag enum class for simplicity and clarity.

* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::layoutBlock):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::RenderObject):
(WebCore::RenderObject::clearNeedsLayout):
(WebCore::RenderObject::setPreferredLogicalWidthsDirty):
(WebCore::RenderObject::invalidateContainerPreferredLogicalWidths):
(WebCore::RenderObject::destroy):
(WebCore::RenderObject::ensureRareData):
(WebCore::RenderObject::removeRareData):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::beingDestroyed const):
(WebCore::RenderObject::everHadLayout const):
(WebCore::RenderObject::childrenInline const):
(WebCore::RenderObject::setChildrenInline):
(WebCore::RenderObject::isAnonymous const):
(WebCore::RenderObject::isFloating const):
(WebCore::RenderObject::isRenderText const):
(WebCore::RenderObject::isRenderBox const):
(WebCore::RenderObject::isInline const):
(WebCore::RenderObject::isReplacedOrInlineBlock const):
(WebCore::RenderObject::isHorizontalWritingMode const):
(WebCore::RenderObject::hasReflection const):
(WebCore::RenderObject::isRenderFragmentedFlow const):
(WebCore::RenderObject::hasOutlineAutoAncestor const):
(WebCore::RenderObject::paintContainmentApplies const):
(WebCore::RenderObject::hasSVGTransform const):
(WebCore::RenderObject::isExcludedFromNormalLayout const):
(WebCore::RenderObject::setIsExcludedFromNormalLayout):
(WebCore::RenderObject::hasLayer const):
(WebCore::RenderObject::selfNeedsLayout const):
(WebCore::RenderObject::needsPositionedMovementLayout const):
(WebCore::RenderObject::posChildNeedsLayout const):
(WebCore::RenderObject::needsSimplifiedNormalFlowLayout const):
(WebCore::RenderObject::normalChildNeedsLayout const):
(WebCore::RenderObject::preferredLogicalWidthsDirty const):
(WebCore::RenderObject::hasNonVisibleOverflow const):
(WebCore::RenderObject::hasTransformRelatedProperty const):
(WebCore::RenderObject::setFloating):
(WebCore::RenderObject::setInline):
(WebCore::RenderObject::setIsRenderText):
(WebCore::RenderObject::setIsRenderBox):
(WebCore::RenderObject::setReplacedOrInlineBlock):
(WebCore::RenderObject::setHorizontalWritingMode):
(WebCore::RenderObject::setHasNonVisibleOverflow):
(WebCore::RenderObject::setHasLayer):
(WebCore::RenderObject::setHasTransformRelatedProperty):
(WebCore::RenderObject::setNeedsPositionedMovementLayoutBit):
(WebCore::RenderObject::setNormalChildNeedsLayoutBit):
(WebCore::RenderObject::setPosChildNeedsLayoutBit):
(WebCore::RenderObject::setNeedsSimplifiedNormalFlowLayoutBit):
(WebCore::RenderObject::setEverHadLayout):
(WebCore::RenderObject::hasRareData const):
(WebCore::RenderObject::RenderObjectBitfields::flags const):
(WebCore::RenderObject::RenderObjectBitfields::hasFlag const):
(WebCore::RenderObject::RenderObjectBitfields::setFlag):
(WebCore::RenderObject::RenderObjectBitfields::clearFlag):
(WebCore::RenderObject::setNeedsLayout):
(WebCore::RenderObject::needsSimplifiedNormalFlowLayoutOnly const):
(WebCore::RenderObject::needsLayout const):
(WebCore::RenderObject::needsPositionedMovementLayoutOnly const):
(WebCore::RenderObject::setHasRareData): Deleted.
(WebCore::RenderObject::RenderObjectBitfields::RenderObjectBitfields): Deleted.

Canonical link: <a href="https://commits.webkit.org/272255@main">https://commits.webkit.org/272255@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9cab6586d43c26dfda7f4efdee5f1b280d7852ac

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/31126 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/9798 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/32814 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/33632 "Built successfully") | [❌ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/28094 "Hash 9cab6586 for PR 21996 does not build (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/31889 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/12142 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/7057 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/27928 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/31458 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/8256 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/27820 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/7093 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/7262 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/27715 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/34971 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/28324 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/28172 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/33399 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/7302 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/5356 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/31237 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/8990 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8005 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4037 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/7837 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderObject.h:
```diff
@@ -1120,6 +1119,31 @@ class RenderObject : public CachedImageClient, public CanMakeCheckedPtr {
+        HasNonVisibleOverflow = 1 << 18,
+        HasTransformRelatedProperty = 1 << 19,
+        ChildrenInline = 1 << 20,
+        // 3 bits left
```
This is a bit confusing since it's a `uint32_t`. Will trying to use bit 23 cause a compile-time failure?

rniwa replied:
Unfortunately not. But that'll increase the size of RenderObjectBitfields so hopefully we'd check that there.

rniwa replied:
Actually, I should add an assertion in setFlag that m_flags is wide enough for the flag. Will do that.

rniwa commented on Source/WebCore/rendering/RenderObject.h:
```diff
@@ -1128,74 +1152,19 @@ class RenderObject : public CachedImageClient, public CanMakeCheckedPtr {
-            , m_isAnonymous(node.isDocumentNode())
-            , m_isRenderText(false)
-            , m_isRenderBox(false)
-            , m_isInline(true)
```
This defaults to true. I'm flipping the flag to be IsBlock instead so that the default value can be 0.

rniwa committed:
Replace 1-bit bitfields in RenderObjectBitfields with RenderObjectFlag
https://bugs.webkit.org/show_bug.cgi?id=266609

Reviewed by Simon Fraser.

Replace macros to generate 1-bit bitfields in RenderObjectBitfields with newly introduced
RenderObjectFlag enum class for simplicity and clarity.

* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::layoutBlock):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::RenderObject):
(WebCore::RenderObject::clearNeedsLayout):
(WebCore::RenderObject::setPreferredLogicalWidthsDirty):
(WebCore::RenderObject::invalidateContainerPreferredLogicalWidths):
(WebCore::RenderObject::destroy):
(WebCore::RenderObject::ensureRareData):
(WebCore::RenderObject::removeRareData):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::beingDestroyed const):
(WebCore::RenderObject::everHadLayout const):
(WebCore::RenderObject::childrenInline const):
(WebCore::RenderObject::setChildrenInline):
(WebCore::RenderObject::isAnonymous const):
(WebCore::RenderObject::isFloating const):
(WebCore::RenderObject::isRenderText const):
(WebCore::RenderObject::isRenderBox const):
(WebCore::RenderObject::isInline const):
(WebCore::RenderObject::isReplacedOrInlineBlock const):
(WebCore::RenderObject::isHorizontalWritingMode const):
(WebCore::RenderObject::hasReflection const):
(WebCore::RenderObject::isRenderFragmentedFlow const):
(WebCore::RenderObject::hasOutlineAutoAncestor const):
(WebCore::RenderObject::paintContainmentApplies const):
(WebCore::RenderObject::hasSVGTransform const):
(WebCore::RenderObject::isExcludedFromNormalLayout const):
(WebCore::RenderObject::setIsExcludedFromNormalLayout):
(WebCore::RenderObject::hasLayer const):
(WebCore::RenderObject::selfNeedsLayout const):
(WebCore::RenderObject::needsPositionedMovementLayout const):
(WebCore::RenderObject::posChildNeedsLayout const):
(WebCore::RenderObject::needsSimplifiedNormalFlowLayout const):
(WebCore::RenderObject::normalChildNeedsLayout const):
(WebCore::RenderObject::preferredLogicalWidthsDirty const):
(WebCore::RenderObject::hasNonVisibleOverflow const):
(WebCore::RenderObject::hasTransformRelatedProperty const):
(WebCore::RenderObject::setFloating):
(WebCore::RenderObject::setInline):
(WebCore::RenderObject::setIsRenderText):
(WebCore::RenderObject::setIsRenderBox):
(WebCore::RenderObject::setReplacedOrInlineBlock):
(WebCore::RenderObject::setHorizontalWritingMode):
(WebCore::RenderObject::setHasNonVisibleOverflow):
(WebCore::RenderObject::setHasLayer):
(WebCore::RenderObject::setHasTransformRelatedProperty):
(WebCore::RenderObject::setNeedsPositionedMovementLayoutBit):
(WebCore::RenderObject::setNormalChildNeedsLayoutBit):
(WebCore::RenderObject::setPosChildNeedsLayoutBit):
(WebCore::RenderObject::setNeedsSimplifiedNormalFlowLayoutBit):
(WebCore::RenderObject::setEverHadLayout):
(WebCore::RenderObject::hasRareData const):
(WebCore::RenderObject::RenderObjectBitfields::flags const):
(WebCore::RenderObject::RenderObjectBitfields::hasFlag const):
(WebCore::RenderObject::RenderObjectBitfields::setFlag):
(WebCore::RenderObject::RenderObjectBitfields::clearFlag):
(WebCore::RenderObject::setNeedsLayout):
(WebCore::RenderObject::needsSimplifiedNormalFlowLayoutOnly const):
(WebCore::RenderObject::needsLayout const):
(WebCore::RenderObject::needsPositionedMovementLayoutOnly const):
(WebCore::RenderObject::setHasRareData): Deleted.
(WebCore::RenderObject::RenderObjectBitfields::RenderObjectBitfields): Deleted.

Canonical link: https://commits.webkit.org/272255@main
